### PR TITLE
[SPARK-21589][SQL][DOC] Add documents about Hive UDF/UDTF/UDAF

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added documents about unsupported functions in Hive UDF/UDTF/UDAF.
This pr relates to #18768 and #18527.

## How was this patch tested?
N/A

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile If you get time, could you check this? Thanks!

File: docs/sql-programming-guide.md [hunk 1]
```diff
@@ -1903,6 +1903,25 @@ releases of Spark SQL.
   Hive can optionally merge the small files into fewer large files to avoid overflowing the HDFS
   metadata. Spark SQL does not support that.
 
+**Hive UDF/UDTF/UDAF**
+
+Spark SQL implements the basic functionality of the Hive UDF/UDTF/UDAF, but does not support all the APIs for users.
+Some of them are meaningless in Spark and the others are rarely used by users.
+Below is a list of major APIs we don't support in Spark SQL:
+
+* `getRequiredJars` and `getRequiredFiles` (`UDF` and `GenericUDF`) are functions to to automatically
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: `to to` -> `to`

File: docs/sql-programming-guide.md [hunk 2]
```diff
@@ -1903,6 +1903,25 @@ releases of Spark SQL.
...
+  include additional resources required by this UDF.
+* `initialize(StructObjectInspector)` in `GenericUDTF` is not supported yet. Spark SQL currently uses
+  a deprecated interface `initialize(ObjectInspector[])` only.
+* `configure` (`GenericUDF`, `GenericUDTF`, and `GenericUDAFEvaluator`) is a function to initialize
+  functions with `MapredContext`. But, Spark SQL does not use `MapredContext` internally.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: > functions with `MapredContext`, which is inapplicable to Spark.

File: docs/sql-programming-guide.md [hunk 3]
```diff
@@ -1903,6 +1903,25 @@ releases of Spark SQL.
...
+* `close` (`GenericUDF` and `GenericUDAFEvaluator`) is a function to release associated resources.
+  Spark SQL does not call this function when tasks finished.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: `finished` -> `finish`

File: docs/sql-programming-guide.md [hunk 4]
```diff
@@ -1903,6 +1903,25 @@ releases of Spark SQL.
...
+* `reset` (`GenericUDAFEvaluator`) is a function to re-initialize aggregation for reusing the same aggregation.
+  Spark SQL currently does not support the reuse of aggregation.
+* `getWindowingEvaluator` (`GenericUDAFEvaluator`) is a function to optimize aggregation by evaluating
+  an aggregate over a fixed window. Spark SQL does not support this optimization yet.
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: Please remove ` Spark SQL does not support this optimization yet`

File: docs/sql-programming-guide.md [hunk 5]
```diff
@@ -1903,6 +1903,25 @@ releases of Spark SQL.
   Hive can optionally merge the small files into fewer large files to avoid overflowing the HDFS
   metadata. Spark SQL does not support that.
 
+**Hive UDF/UDTF/UDAF**
+
+Spark SQL implements the basic functionality of the Hive UDF/UDTF/UDAF, but does not support all the APIs for users.
+Some of them are meaningless in Spark and the others are rarely used by users.
+Below is a list of major APIs we don't support in Spark SQL:
```
Comment (gatorsmile) [inline docs/sql-programming-guide.md]: `we don't support in Spark SQL:` -> `that are not supported by Spark SQL:`

**Comment by gatorsmile** [comment]
Thanks for working on it! Just left some minor comments.

**Comment by maropu** [comment]
@gatorsmile ok, fixed.

File: docs/sql-programming-guide.md [hunk 6]
```diff
@@ -1903,6 +1903,23 @@ releases of Spark SQL.
...
+**Hive UDF/UDTF/UDAF**
+
+Not all the APIs of the Hive UDF/UDTF/UDAF are supported by Spark SQL. Below are the unsupported APIs:
+
+* `getRequiredJars` and `getRequiredFiles` (`UDF` and `GenericUDF`) are functions to automatically
+  include additional resources required by this UDF.
+* `initialize(StructObjectInspector)` in `GenericUDTF` is not supported yet. Spark SQL currently uses
+  a deprecated interface `initialize(ObjectInspector[])` only.
+* `configure` (`GenericUDF`, `GenericUDTF`, and `GenericUDAFEvaluator`) is a function to initialize
+  functions with `MapredContext`, which is inapplicable to Spark. But, Spark SQL does not use `MapredContext` internally.
```
Comment (viirya) [inline docs/sql-programming-guide.md]: nit: `But` looks redundant here, because there's `inapplicable` before. Looks like negative to negative...

Comment (maropu) [inline docs/sql-programming-guide.md]: removed. Thanks!

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Pull request closed without merging**